### PR TITLE
[CSL-1676] Send keep-alive like packets after relay subscription

### PR DESCRIPTION
**Pull request by arybczak**
That should take care of [CSL-1676](https://issues.serokell.io/issue/CSL-1676).

The only issue is that I'm not sure how to test that change  - running `demo.sh` doesn't trigger subscription worker on any of the nodes.

### CODE REVIEW

File: infra/Pos/Subscription/Common.hs [hunk 1]
```diff
@@ -63,10 +65,14 @@ subscribeTo
...
     logNotice $ msgSubscribingTo peer
-    outcome <- try $ withConnectionTo sendActions peer $ \_peerData ->
-           pure $ Conversation $ \conv -> do
-               send conv MsgSubscribe
-               recv conv 0 :: m (Maybe Void) -- Other side will never send
+    outcome <- try $ withConnectionTo sendActions peer $ \_peerData -> do
+        pure $ Conversation $ \(conv :: ConversationActions MsgSubscribe Void m) -> do
+            send conv MsgSubscribe
+            forever $ do
+                threadDelay (30 :: Second)
```
Comment (dcoutts) [inline infra/Pos/Subscription/Common.hs]: Ideally we can do better than this. Perhaps it's too complex, but I'd like us to check.

Ideally we delay 'til 30s after the last block was received. Or equivalently, we wait 30s, but reset/postpone our timer each time we receive a block. It doesn't matter if it's easier for us to use receiving a block header as the trigger, or receive anything from the target peer, or applying a block, whatever is easier.

File: infra/Pos/Subscription/Common.hs [hunk 2]
```diff
@@ -88,14 +94,18 @@ subscriptionListener
...
+    whenJust mbMsg $ \case
+        MsgSubscribe -> do
+            let peers = simplePeers [(nodeType, nodeId)]
+            bracket
+              (updatePeersBucket BucketSubscriptionListener (<> peers))
+              (\added -> when added $
+                void $ updatePeersBucket BucketSubscriptionListener (removePeer nodeId))
+              (\added -> when added $
+                void $ recvLimited conv) -- if not added, close the conversation
+        MsgSubscribeKeepAlive -> logDebug $ sformat
```
Comment (dcoutts) [inline infra/Pos/Subscription/Common.hs]: This is not quite right. The `MsgSubscribeKeepAlive` at the top level here is for opening a conversation with that message, which in this case would always be a protocol error. The case where we expect to receive a keep-alive is in the `void $ recvLimited conv` line above. Previously that was doing a `recvLimited` in the expectation that no message would ever be received, and thereby keeping the connection open, now we need a loop that received and discards keep alive messages. Something like:

```
 -          void $ recvLimited conv) -- if not added, close the conversation
 +          forever $ do
 +            Just MsgSubscribeKeepAlive <- recvLimited conv
 +            logDebug ...
```
Ok, `forever` and the do style pattern matching is terrible error handling, but you get the idea.

**Comment by avieth** [comment]
> The only issue is that I'm not sure how to test that change - running demo.sh doesn't trigger subscription worker on any of the nodes.

Right, maybe the demo should always launch a relay and an edge node. There's a variation on `demo.sh` that brings up 3 cores, a relay, an edge, and a transaction generator: `scripts/bench/runbench.sh`.

File: infra/Pos/Util/Timer.hs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+waitTimer Timer{..} = do
+  done <- readTVar =<< readTVar delaySemaphore
+  unless done retry
+
+-- Start the timer. If the function is called before @t + n@, where @t@ is the
+-- time startTimer was called for the last time and @n@ is the number of
+-- microseconds associated with the Timer, the timer is restarted.
+startTimer :: MonadIO m => Timer -> m ()
+startTimer Timer{..} = liftIO (registerDelay delayDuration)
+    >>= atomically . writeTVar delaySemaphore
```
Comment (dcoutts) [inline infra/Pos/Util/Timer.hs]: Very nice.

**Comment by dcoutts** [comment]
@arybczak can you update #1853 now. No need to fix the TODO for the 1.0 branch, just for master.

**Comment by avieth** [comment]
> For the master version here, @avieth can you point out to @arybczak where to get the slot duration from so we can fix that TODO.

`getCurrentEpochSlotDuration` from `infra/Pos/Slotting/Util.hs`.

**Pull request merged**